File: tools/packages/fetch-client/packages_py/fetch_client/src/fetch_client/health/status_checker.py
"""
Fetch Status Checker Implementation.
"""
import time
from typing import Optional, Any, Dict
import httpx

from fetch_client.client import FetchClient
from fetch_client.config import AuthConfig as FetchAuthConfig, ClientConfig

from .models import FetchStatus, FetchStatusResult

class FetchStatusChecker:
    """Check fetch connectivity to a provider using runtime config."""

    # Default health endpoints for provider types
    DEFAULT_ENDPOINTS = {
        "gemini_openai": "/models",
        "openai": "/models",
        "anthropic": "/models",
        "github": "/user",
        "default": "/",
    }

    def __init__(
        self,
        provider_name: str,
        runtime_config: Any,  # ComputeAllResult
        timeout_seconds: float = 10.0,
        endpoint_override: Optional[str] = None,
    ):
        self.provider_name = provider_name
        self.runtime_config = runtime_config
        self.timeout_seconds = timeout_seconds
        self.endpoint_override = endpoint_override

    async def check(self) -> FetchStatusResult:
        """Execute health check and return result."""
        from datetime import datetime, timezone

        timestamp = datetime.now(timezone.utc).isoformat()
        start_time = time.perf_counter()

        # Validate config
        if not self.runtime_config.config.get("base_url"):
            return self._config_error("base_url is required", timestamp)

        health_endpoint = self._resolve_health_endpoint()
        base_url = self.runtime_config.config.get("base_url")
        # Ensure clean url construction
        full_url = f"{base_url.rstrip('/')}/{health_endpoint.lstrip('/')}"

        request_info = {
            "method": "GET",
            "url": full_url,
            "timeout_seconds": self.timeout_seconds,
        }

        config_used = self._build_config_used(health_endpoint)

        client = None
        try:
            client = self._create_client()
            response = await client.get(
                health_endpoint,
                params=None # Could allow params override if needed
            )
            latency_ms = (time.perf_counter() - start_time) * 1000

            return FetchStatusResult(
                provider_name=self.provider_name,
                status=self._status_from_code(response.status),
                latency_ms=round(latency_ms, 2),
                timestamp=timestamp,
                request=request_info,
                response={
                    "status_code": response.status,
                    "status_text": response.status_text,
                    "content_type": response.headers.get("content-type"),
                    "body_preview": self._truncate(str(response.data)),
                },
                config_used=config_used,
            )

        except httpx.TimeoutException as e:
            latency_ms = (time.perf_counter() - start_time) * 1000
            return FetchStatusResult(
                provider_name=self.provider_name,
                status=FetchStatus.TIMEOUT,
                latency_ms=round(latency_ms, 2),
                timestamp=timestamp,
                request=request_info,
                config_used=config_used,
                error={"type": "TimeoutException", "message": str(e)},
            )

        except httpx.ConnectError as e:
            latency_ms = (time.perf_counter() - start_time) * 1000
            return FetchStatusResult(
                provider_name=self.provider_name,
                status=FetchStatus.CONNECTION_ERROR,
                latency_ms=round(latency_ms, 2),
                timestamp=timestamp,
                request=request_info,
                config_used=config_used,
                error={"type": "ConnectError", "message": str(e)},
            )

        except Exception as e:
            latency_ms = (time.perf_counter() - start_time) * 1000
            return FetchStatusResult(
                provider_name=self.provider_name,
                status=FetchStatus.ERROR,
                latency_ms=round(latency_ms, 2),
                timestamp=timestamp,
                request=request_info,
                config_used=config_used,
                error={
                    "type": type(e).__name__,
                    "message": str(e),
                },
            )
        finally:
            if client:
                await client.close()

    def _resolve_health_endpoint(self) -> str:
        """Resolve the health endpoint to use."""
        if self.endpoint_override:
            return self.endpoint_override
        if self.runtime_config.config.get("health_endpoint"):
            return self.runtime_config.config["health_endpoint"]
        return self.DEFAULT_ENDPOINTS.get(
            self.provider_name,
            self.DEFAULT_ENDPOINTS["default"]
        )

    def _create_client(self) -> FetchClient:
        """Create fetch client from runtime config."""
        
        auth_config = None
        if self.runtime_config.auth_config:
            auth_config = self._convert_auth_config(self.runtime_config.auth_config)

        # Build ClientConfig
        config = ClientConfig(
            base_url=self.runtime_config.config["base_url"],
            auth=auth_config,
            headers=self.runtime_config.config.get("headers", {}),
            timeout=self.timeout_seconds,
        )
        return FetchClient.create(config)

    def _convert_auth_config(self, auth: Any) -> FetchAuthConfig:
        """Convert yaml_config AuthConfig to fetch_client AuthConfig."""
        # Note: 'auth' here is config_loader.AuthConfig or similar, containing resolved secrets.
        # We handle Pydantic models or plain objects.
        
        type_val = auth.type.value if hasattr(auth.type, 'value') else auth.type
        
        return FetchAuthConfig(
            type=type_val,
            raw_api_key=auth.token,
            username=auth.username,
            password=auth.password,
            email=auth.email,
            header_name=auth.header_name,
        )

    def _build_config_used(self, health_endpoint: str) -> Dict[str, Any]:
        """Build safe config summary."""
        auth = self.runtime_config.auth_config
        
        auth_type_str = None
        if auth:
            auth_type_str = auth.type.value if hasattr(auth.type, 'value') else str(auth.type)

        return {
            "base_url": self.runtime_config.config.get("base_url"),
            "health_endpoint": health_endpoint,
            "auth_type": auth_type_str,
            "auth_resolved": bool(auth and auth.token),
            "is_placeholder": auth.resolution.is_placeholder if auth and auth.resolution else None,
            "proxy_url": self.runtime_config.proxy_config.proxy_url if self.runtime_config.proxy_config else None,
            "headers_count": len(self.runtime_config.config.get("headers", {})),
        }

    def _status_from_code(self, status_code: int) -> FetchStatus:
        """Map HTTP status code to FetchStatus."""
        if 200 <= status_code < 300:
            return FetchStatus.CONNECTED
        elif status_code in (401, 403):
            return FetchStatus.UNAUTHORIZED
        elif 400 <= status_code < 500:
            return FetchStatus.CLIENT_ERROR
        elif status_code >= 500:
            return FetchStatus.SERVER_ERROR
        return FetchStatus.ERROR

    def _config_error(self, message: str, timestamp: str) -> FetchStatusResult:
        """Create config error result."""
        return FetchStatusResult(
            provider_name=self.provider_name,
            status=FetchStatus.CONFIG_ERROR,
            latency_ms=0,
            timestamp=timestamp,
            error={"type": "ConfigError", "message": message},
        )

    @staticmethod
    def _truncate(text: str, max_len: int = 500) -> str:
        """Truncate text for preview."""
        if len(text) <= max_len:
            return text
        return text[:max_len] + "... (truncated)"

File: tools/packages/fetch-client/packages_mjs/fetch-client/src/health/status-checker.ts
/**
 * Fetch Status Checker Implementation.
 */
import { FetchClient } from "../client.js";
import { AuthConfig, ClientConfig } from "../config.js";
import { FetchStatus, FetchStatusResult } from "./models.js";

interface ComputeAllResult {
    config: Record<string, any>;
    auth_config?: any; // AppYamlConfig auth config structure
    proxy_config?: any;
    network_config?: any;
    headers?: Record<string, string>;
}

export class FetchStatusChecker {
    private static DEFAULT_ENDPOINTS: Record<string, string> = {
        gemini_openai: "/models",
        openai: "/models",
        anthropic: "/models",
        github: "/user",
        default: "/",
    };

    constructor(
        private provider_name: string,
        private runtime_config: ComputeAllResult,
        private timeout_seconds: number = 10.0,
        private endpoint_override?: string
    ) { }

    public async check(): Promise<FetchStatusResult> {
        const timestamp = new Date().toISOString();
        const startTime = performance.now();

        // Validate config
        if (!this.runtime_config.config.base_url) {
            return this._configError("base_url is required", timestamp);
        }

        const healthEndpoint = this._resolveHealthEndpoint();
        const baseUrl = this.runtime_config.config.base_url;
        // Basic url joining to look nice in status
        const fullUrl = `${baseUrl.replace(/\/$/, "")}/${healthEndpoint.replace(/^\//, "")}`;

        const requestInfo = {
            method: "GET",
            url: fullUrl,
            timeout_seconds: this.timeout_seconds,
        };

        const configUsed = this._buildConfigUsed(healthEndpoint);

        let client: FetchClient | undefined;

        try {
            client = this._createClient();
            const response = await client.get(healthEndpoint);

            const latencyMs = performance.now() - startTime;

            // Read body safely
            let bodyPreview = "";
            try {
                const text = response.data ? (typeof response.data === 'string' ? response.data : JSON.stringify(response.data)) : "";
                bodyPreview = this._truncate(text);
            } catch (e) {
                bodyPreview = "[Unable to read body]";
            }

            return {
                provider_name: this.provider_name,
                status: this._statusFromCode(response.status),
                latency_ms: Number(latencyMs.toFixed(2)),
                timestamp,
                request: requestInfo,
                response: {
                    status_code: response.status,
                    status_text: response.statusText,
                    content_type: response.headers instanceof Map ? response.headers.get("content-type") : (response.headers as any)["content-type"],
                    body_preview: bodyPreview,
                },
                config_used: configUsed,
            };

        } catch (e: any) {
            const latencyMs = performance.now() - startTime;

            let status = FetchStatus.ERROR;
            let errorType = "Error";
            let errorMessage = String(e);

            // Handle specific error types if possible (undici errors)
            if (e.name === "ConnectTimeoutError" || e.code === "UND_ERR_CONNECT_TIMEOUT") {
                status = FetchStatus.TIMEOUT;
                errorType = "TimeoutException";
            } else if (e.name === "SocketError" || e.code === "UND_ERR_SOCKET") {
                status = FetchStatus.CONNECTION_ERROR;
                errorType = "ConnectError";
            }

            return {
                provider_name: this.provider_name,
                status,
                latency_ms: Number(latencyMs.toFixed(2)),
                timestamp,
                request: requestInfo,
                config_used: configUsed,
                error: {
                    type: errorType,
                    message: errorMessage,
                    code: e.code,
                    name: e.name
                },
            };

        } finally {
            // BaseClient/undici dispatcher management is handled by the instance.
            // We should ideally close it if we created a dedicated one, but FetchClient
            // doesn't expose close() primarily because it shares the global dispatcher if not specified.
            // However, BaseClient DOES have a close() method if we want to be clean.
            // checking if we can cast or if we exposed it. BaseClient has close().
            if (client && typeof (client as any).close === 'function') {
                await (client as any).close();
            }
        }
    }

    private _resolveHealthEndpoint(): string {
        if (this.endpoint_override) return this.endpoint_override;
        if (this.runtime_config.config.health_endpoint) {
            return this.runtime_config.config.health_endpoint;
        }
        return FetchStatusChecker.DEFAULT_ENDPOINTS[this.provider_name] || FetchStatusChecker.DEFAULT_ENDPOINTS["default"];
    }

    private _createClient(): FetchClient {
        let authConfig: AuthConfig | undefined;
        if (this.runtime_config.auth_config) {
            authConfig = this._convertAuthConfig(this.runtime_config.auth_config);
        }

        const config: ClientConfig = {
            baseUrl: this.runtime_config.config.base_url,
            auth: authConfig,
            contentType: "application/json",
            headers: this.runtime_config.config.headers || {},
            timeout: this.timeout_seconds * 1000, // Client expects ms
        };

        return FetchClient.create(config);
    }

    private _convertAuthConfig(auth: any): AuthConfig {
        // Convert yaml_config/AppYamlConfig auth structure to FetchClient AuthConfig
        // Note: auth.type might be an Enum object or string depending on source
        const typeVal = auth.type?.value || auth.type;

        return {
            type: asAuthType(typeVal),
            rawApiKey: auth.token,
            username: auth.username,
            password: auth.password,
            email: auth.email,
            headerName: auth.header_name,
        };
    }

    private _buildConfigUsed(healthEndpoint: string): Record<string, any> {
        const auth = this.runtime_config.auth_config;
        const authTypeStr = auth?.type?.value || auth?.type;

        return {
            base_url: this.runtime_config.config.base_url,
            health_endpoint: healthEndpoint,
            auth_type: authTypeStr,
            auth_resolved: Boolean(auth && auth.token),
            is_placeholder: auth?.resolution?.is_placeholder,
            proxy_url: this.runtime_config.proxy_config?.proxy_url,
            headers_count: Object.keys(this.runtime_config.config.headers || {}).length,
        };
    }

    private _statusFromCode(statusCode: number): FetchStatus {
        if (statusCode >= 200 && statusCode < 300) return FetchStatus.CONNECTED;
        if (statusCode === 401 || statusCode === 403) return FetchStatus.UNAUTHORIZED;
        if (statusCode >= 400 && statusCode < 500) return FetchStatus.CLIENT_ERROR;
        if (statusCode >= 500) return FetchStatus.SERVER_ERROR;
        return FetchStatus.ERROR;
    }

    private _configError(message: string, timestamp: string): FetchStatusResult {
        return {
            provider_name: this.provider_name,
            status: FetchStatus.CONFIG_ERROR,
            latency_ms: 0,
            timestamp,
            error: { type: "ConfigError", message },
        };
    }

    private _truncate(text: string, maxLen: number = 500): string {
        if (text.length <= maxLen) return text;
        return text.substring(0, maxLen) + "... (truncated)";
    }
}

// Helper to safely cast auth type string
import { AuthType } from "../types.js";
function asAuthType(val: string): AuthType {
    // In a real scenario we might validate, but for now we trust the upstream resolver or default to custom
    return val as AuthType;
}
